aten/src/ATen/native/ReduceOps.cpp

```diff
@@ -261,6 +261,23 @@ static inline Tensor &mean_out(Tensor &result, const Tensor &self, IntList dim,
       toString(scalarType),
       " instead.");
   ScalarType dtype = get_dtype(result, self, opt_dtype, true);
+  // TODO: the TensorIterator reduction implementation of mean
+  // (mean_kernel_impl()) is unvectorized and leads to very poor performance
+  // for production workloads. Once that's fixed, the following code can be used
+  // in lieu of the sum + divide implementation below.
+  if (!self.is_cuda() && self.is_hip()) {
```

Comment:
I'm confused. The issue you put out says that this is slow on CPUs, but the implementation you have here checks that `self` is on an AMD GPU?

Reply:
This looks like a bug introduced while trying to get the tests to pass. It is probably safer to make this a positive check `self.is_cpu()` rather than a negative check.

Reply:
`is_cpu()` doesn't seem to exist, that's why I didnt use it:

```
caffe2/aten/src/ATen/native/ReduceOps.cpp: In function 'at::Tensor& at::native::mean_out(at::Tensor&, const at::Tensor&, c10::IntList, bool, c10::optional<c10::ScalarType>)':
caffe2/aten/src/ATen/native/ReduceOps.cpp:268:12: error: 'const class at::Tensor' has no member named 'is_cpu'
At global scope:
cc1plus: error: unrecognized command line option '-Wno-class-memaccess' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-stringop-overflow' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-braced-scalar-init' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-absolute-value' [-Werror]
cc1plus: all warnings being treated as errors
Parsing buck files: finished in 1.1 sec
Building: finished in 19.7 sec (100%) 14429/14431 jobs, 1 updated
  Total time: 20.9 sec
More details at https://our.intern.facebook.com/intern/buck/build/a72c7672-d6ce-4b63-ab70-b0cc8cc
Command failed with exit code 1.
stderr: caffe2/aten/src/ATen/native/ReduceOps.cpp: In function 'at::Tensor& at::native::mean_out(at::Tensor&, const at::Tensor&, c10::IntList, bool, c10::optional<c10::ScalarType>)':
caffe2/aten/src/ATen/native/ReduceOps.cpp:268:12: error: 'const class at::Tensor' has no member named 'is_cpu'
At global scope:
cc1plus: error: unrecognized command line option '-Wno-class-memaccess' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-stringop-overflow' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-braced-scalar-init' [-Werror]
cc1plus: error: unrecognized command line option '-Wno-absolute-value' [-Werror]
cc1plus: all warnings being treated as errors
    When running <c++ preprocess_and_compile>.
    When building rule //caffe2/aten:ATen-cpu#compile-ReduceOps.cpp.oe29fdf65,gcc-5-glibc-2.23-gcc
```
Can anyone advise on this? It seems like something we should have

This is indeed a new bug we've introduced, the is_hip call should have a negation.

Reply:
`self.device().is_cpu()` seems to exist